netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java

```diff
@@ -0,0 +1,305 @@
+    if (stream != null) {
+      cancelPendingStream(stream);
+      promise.setSuccess();
+    } else {
+      promise.setFailure(connectionError(PROTOCOL_ERROR, "Stream does not exist %d", streamId));
+    }
+    return promise;
+  }
+
+  @Override
+  public ChannelFuture writeData(ChannelHandlerContext ctx, int streamId, ByteBuf data, int padding,
+                                 boolean endOfStream, ChannelPromise promise) {
+    if (existingStream(streamId)) {
+      return super.writeData(ctx, streamId, data, padding, endOfStream, promise);
+    }
+    PendingStream pendingStream = pendingStreams.get(streamId);
+    if (pendingStream != null) {
+      pendingStream.frames.add(new DataFrame(data, padding, endOfStream, promise));
+    } else {
+      ReferenceCountUtil.safeRelease(data);
+      promise.setFailure(connectionError(PROTOCOL_ERROR, "Stream does not exist %d", streamId));
+    }
+    return promise;
+  }
+
+  @Override
+  public ChannelFuture writeSettingsAck(ChannelHandlerContext ctx, ChannelPromise promise) {
+    ChannelFuture future = super.writeSettingsAck(ctx, promise);
+    // After having received a SETTINGS frame, the maximum number of concurrent streams
+    // might have changed. So try to create some buffered streams.
+    tryCreatePendingStreams();
+    return future;
+  }
+
+  @Override
+  public void close() {
+    super.close();
+    cancelPendingStreams();
+  }
+
+  private void tryCreatePendingStreams() {
+    while (!pendingStreams.isEmpty() && connection().local().acceptingNewStreams()) {
+      createNextPendingStream();
+    }
+  }
+
+  private void cancelPendingStreams() {
+    while (!pendingStreams.isEmpty()) {
+      PendingStream stream = pendingStreams.pollFirstEntry().getValue();
+      cancelPendingStream(stream);
+    }
+  }
+
+  private void createNextPendingStream() {
+    Map.Entry<Integer, PendingStream> entry = pendingStreams.pollFirstEntry();
+    if (entry == null) {
+      return;
+    }
+    PendingStream pendingStream = entry.getValue();
+    for (Frame frame : pendingStream.frames) {
+      if (frame instanceof HeadersFrame) {
+        HeadersFrame headers = (HeadersFrame) frame;
+        super.writeHeaders(pendingStream.ctx, pendingStream.streamId, headers.headers,
+                headers.streamDependency, headers.weight, headers.exclusive,
+                headers.padding, headers.endOfStream, headers.promise);
+      } else {
+        DataFrame data = (DataFrame) frame;
+        super.writeData(pendingStream.ctx, pendingStream.streamId, data.data,
+                data.padding, data.endOfStream, data.promise);
+      }
+    }
+    largestCreatedStreamId = pendingStream.streamId;
+  }
+
+  private static void cancelPendingStream(PendingStream stream) {
+    for (Frame frame : stream.frames) {
+      frame.fail("Buffered stream canceled.");
+    }
+  }
+
+  private boolean existingStream(int streamId) {
+    return streamId <= largestCreatedStreamId;
```

Comment:
I think you also need to check if the stream was created by the local endpoint (even/odd). I think there are methods in the `Http2Connection` interface that might help.